Alerting: Create endpoints for exporting in provisioning file format

JacobsonMT opened the pull request:
Updated: 2022-12-09

**What is this feature?**

This adds provisioning endpoints for downloading alert rules and alert rule groups in a format that is compatible with file provisioning. Each endpoint supports both json and yaml response types via `Accept` header as well as a query parameter `download=true/false` that will set `Content-Disposition` to recommend initiating a download or inline display.

NOTE: `Content-Type` will be `text/yaml` for inline display and `application/yaml` for downloads. This is because the default behaviour in some browsers (at least Chrome/Firefox) is to download `application/yaml` regardless of `Content-Disposition`

**Why do we need this feature?**

This feature is intended to facilitate provisioning file creation for alerts as well as allowing for a feature for downloading provisioning file exports directly from the UI.

**Which issue(s) does this PR fix?**:
Partial: https://github.com/grafana/grafana/issues/48622

**Special notes for your reviewer**:

For Backend Platform:

The only file to really look at is `pkg/api/response/response.go`, if you prefer I can leave this file unmodified and move the code to the alerting provisioning api.

Otherwise:

The structure of the endpoints URLs is very much up for discussion. Currently they are:

`/api/v1/provisioning/alert-rules/export`
`/api/v1/provisioning/alert-rules/{UID}/export`
`/api/v1/provisioning/folder/{FolderUID}/rule-groups/{Group}/export`

However, this poses a few annoyances:
- There is an (albeit unlikely) conflict if a user has an alert rule with uid=`export`
- ~~The `GET /api/v1/provisioning/alert-rules/` endpoint doesn't exist, even though `/api/v1/provisioning/alert-rules/export` now will. We can easily fix this by implementing `GET /api/v1/provisioning/alert-rules/` if wanted, as it will follow closely the code in this PR.~~ Endpoint will be implemented in another PR

There are some alternatives, such as using a query parameter like `format=export` instead of the `export` path variable, or pre-prending to the base path like `/api/v1/provisioning/export/alert-rules`. Each have their own issues.

**Screenshots**:

<details>
  <summary>/api/v1/provisioning/alert-rules/export</summary>

![Screenshot from 2022-11-11 00-49-59](https://user-images.githubusercontent.com/8484471/201272520-9e0cfcf4-9fa7-4891-a1f0-63dc1af751e4.png)

</details>

<details>
  <summary>/api/v1/provisioning/alert-rules/{UID}/export</summary>

![Screenshot from 2022-12-09 13-04-40](https://user-images.githubusercontent.com/8484471/206765275-1efc76e4-d547-4745-9ceb-3acf2a316252.png)


</details>

<details>
  <summary>/api/v1/provisioning/folder/{FolderUID}/rule-groups/{Group}/export</summary>

![Screenshot from 2022-12-09 13-05-02](https://user-images.githubusercontent.com/8484471/206765292-317ecfbf-e4a3-40b4-96f4-d3d3a8700bb5.png)


</details>

JacobsonMT committed:
Implement endpoint to export alert group in provisioning file format

JacobsonMT committed:
Implement endpoint to export alert rule in provisioning file format

JacobsonMT committed:
Implement endpoint to export all alert rule in provisioning file format

Also removes use of SignedInUser for getting folder titles in provisioning.

JacobsonMT committed:
Cleanup redundant export api response code

JacobsonMT committed:
Cleanup export endpoint swagger descriptions

JacobsonMT committed:
Regenerate api

JacobsonMT committed:
linting

JacobsonMT committed:
Fix export filename

alexmobo reviewed (dismissed):
Nice piece of work! There are some authorization tests failing though. I saw some authorization changes, I guess that the reason they are failing. 👍🏽

JacobsonMT committed:
Improve response from yaml marshal error in response.YAML

JacobsonMT commented on pkg/services/ngalert/provisioning/alert_rules.go:
```diff
@@ -351,6 +357,117 @@ func (service *AlertRuleService) deleteRules(ctx context.Context, orgID int64, t
+}
+
+// GetRuleGroupWithTitle returns the alert rule group with folder title.
+func (service *AlertRuleService) GetRuleGroupWithTitle(ctx context.Context, orgID int64, namespaceUID, group string) (AlertRuleGroupWithTitle, error) {
```
I had written that initially, but changed it when I noticed that the rest of the `AlertRuleService`'s methods don't return pointers:

https://github.com/grafana/grafana/blob/4de7a3d861eed4f1037b757ffe324c84866eb444/pkg/services/ngalert/api/api_provisioning.go#L55-L64

I'm inclined to keep it this way for now and we can change them in bulk later if need be.

grobinson-grafana replied:
Sounds good to me!

JacobsonMT committed:
Fix broken authorization_test.go

yuri-tceretian commented on pkg/services/ngalert/api/authorization.go:
```diff
@@ -195,7 +195,10 @@ func (api *API) authorize(method, path string) web.Handler {
+		http.MethodGet + "/api/v1/provisioning/alert-rules/export",
+		http.MethodGet + "/api/v1/provisioning/alert-rules/{UID}/export",
+		http.MethodGet + "/api/v1/provisioning/folder/{FolderUID}/rule-groups/{Group}",
+		http.MethodGet + "/api/v1/provisioning/folder/{FolderUID}/rule-groups/{Group}/export":
```
I think we should consider improving RBAC in provisioning because IMO these endpoints should be available to regular users, not just admins.

grobinson-grafana replied:
Matt did you get a chance to follow up on this?

JacobsonMT replied:
I think it's a good idea, but should probably be done in a separate PR

toha-nordic commented:
@JacobsonMT hi, how does work look for this PR? Would be great to get it merged :)

JacobsonMT commented:
> @JacobsonMT hi, how does work look for this PR? Would be great to get it merged :)

Hi @toha-nordic, there's no big blockers here just haven't had the time to put the finishing touches yet.

I'll have time early next week and would expect it to be merged then :)

alexweav commented on pkg/services/ngalert/api/tooling/definitions/provisioning_alert_rules.go:
```diff
@@ -190,6 +234,48 @@ type AlertRuleGroup struct {
+
+// AlertRuleGroupExport is the provisioned export of models.AlertRuleGroup.
+// swagger:model
+type AlertRuleGroupExport struct {
```
This struct is identical to the struct that file provisioning reads in:
https://github.com/grafana/grafana/blob/a77d95807ca0e85dd2745a2f5ba2795ddef44c49/pkg/services/provisioning/alerting/rules_types.go#L26

Can we use one single definition of this rather than creating it twice?

The `services/provisioning/alerting` package is allowed to depend on ngalert packages  (it already depends on models) we could move it here instead of duplicating.

Same goes for the other structs below.

JacobsonMT replied:
I agree this would be better in theory. I spent some time trying to get this to work but eventually noticed that this would mean using the various `values.{String,Int64,...}Value` in the API. This could:
-  Potentially expose their unmarshalling functionality of expanding os env variables. Now, I think we could do this safely by only allowing their use in responses not in inputs but it still gives me pause as it might not be immediately obvious to future devs that we do such a dangerous operation as env expansion directly in the unmarshaller.
- (less important) We can no longer `omitempty` these fields without some annoying null Pointer "hacks".
 
After some discussion offline, I'm going to keep the separate definitions but instead try to define these structs closer to each other to help reduce likelihood of drift.

alexweav commented on pkg/services/ngalert/api/tooling/definitions/provisioning_alert_rules.go:
```diff
@@ -190,6 +234,48 @@ type AlertRuleGroup struct {
 
+// AlertRuleFileExport is the provisioned export of multiple models.AlertRuleGroup.
+// swagger:model
+type AlertRuleFileExport struct {
+	Groups []AlertRuleGroupExport `json:"groups" yaml:"groups"`
```
Also needs `apiVersion` field which should be set to 1.

alexweav commented on pkg/services/ngalert/api/tooling/definitions/provisioning_alert_rules.go:
```diff
@@ -15,6 +24,19 @@ import (
 //       200: ProvisionedAlertRule
 //       404: description: Not found.
 
+// swagger:route GET /api/v1/provisioning/alert-rules/{UID}/export provisioning stable RouteGetAlertRuleExport
```
We might be able to drop this endpoint. File provisioning only knows how to read one format, and it requires you to provide entire rule groups. There isn't anything that knows how to read the .yaml produced by this endpoint, it'd be a fragment. If you fed it back into file provisioning as-is, it would be rejected.

Same might also go for the GET `.../alert-rules/export` endpoint - it returns a fragment of the rules, but not the group info.

JacobsonMT replied:
Fair point, I think I might prefer to embed it in the full file structure (return `AlertRuleFileExport`) instead of removing the endpoint entirely but I'm ok either way.

JacobsonMT commented on pkg/services/ngalert/api/tooling/definitions/provisioning_alert_rules.go:
```diff
@@ -1,12 +1,21 @@
+// Export all alert rules in provisioning file format.
+//
+//     Responses:
+//       200: AlertRuleFileExport
```
`AlertRuleFileExport` is a valid object in file provisioning (after I add api version). Maybe you meant to comment this on `AlertRuleGroupExport`?

JacobsonMT committed:
Add APIVersion to AlertRuleFileExport

JacobsonMT committed:
Modify provisioning exports to always return the full provisioning file

JacobsonMT committed:
Improve some struct naming

JacobsonMT committed:
Fix api_provisioning_test

JacobsonMT committed:
Organize provisioning file rule structs into new package

This is to keep structs with potential to drift closer together. Eventually, other alerting
file structs should also move into this new package, but the rest require some refactoring
that is out of scope for this PR.

JacobsonMT committed:
Regenerate api

JacobsonMT committed:
Regenerate api

JacobsonMT committed:
regenerate api

JacobsonMT committed:
Regenerate api

JacobsonMT commented:
Rebased onto main

grobinson-grafana commented on pkg/api/response/response.go:
```diff
@@ -187,6 +188,26 @@ func JSONStreaming(status int, body interface{}) StreamingResponse {
+		return Error(http.StatusInternalServerError, "body yaml marshal", err)
+	}
+	// As of now, application/yaml is downloaded by default in chrome regardless of Content-Disposition, so we use text/yaml instead.
+	return Respond(status, b).SetHeader("Content-Type", "text/yaml")
```
Should this be `string(b)` so the response isn't an array of bytes?

```suggestion
	return Respond(status, string(b)).SetHeader("Content-Type", "text/yaml")
```

JacobsonMT replied:
`Respond` handles multiple types (`[]byte`, `string`, and `struct`) so it will correctly handle this here. It might have been cleaner in to modify `Respond` to handle `yaml` marshalling internally as an option but I wanted to avoid modifying existing methods in this file as much as possible since it's used in so many unrelated places.

JacobsonMT replied:
@JohnnyQQQQ 

> LGTM, there is one change that I'm missing and it's about the documentation. We should update the documentation [here](https://github.com/grafana/grafana/blob/db51e963de5cd362cd52d24462b33c237c8f8514/docs/sources/alerting/set-up/provision-alerting-resources/file-provisioning/index.md#L32) to use the new endpoint.

Updated api markdown and docs to use export endpoints here: ac1a3f1be0176d35569c4be33e93b255789221c4 and df5b5a242aa31fbb2481485a4b3feed1d4a55335

grobinson-grafana commented on pkg/api/response/response.go:
```diff
@@ -187,6 +188,26 @@ func JSONStreaming(status int, body interface{}) StreamingResponse {
 
+// JSONDownload creates a JSON response indicating that it should be downloaded.
+func JSONDownload(status int, body interface{}, filename string) *NormalResponse {
+	return JSON(status, body).SetHeader("Content-Disposition", fmt.Sprintf(`attachment;filename="%s"`, filename))
```
Feel free to indent these if you think it might improve readability

```suggestion
	return JSON(status, body).
		SetHeader("Content-Disposition", fmt.Sprintf(`attachment;filename="%s"`, filename))
```

JacobsonMT replied:
Fixed b68338c787c7dc807a33b6635c6ad130e7dcf3dd

grobinson-grafana commented on pkg/api/response/response.go:
```diff
@@ -187,6 +188,26 @@ func JSONStreaming(status int, body interface{}) StreamingResponse {
+
+// YAMLDownload creates a YAML response indicating that it should be downloaded.
+func YAMLDownload(status int, body interface{}, filename string) *NormalResponse {
+	return YAML(status, body).SetHeader("Content-Type", "application/yaml").SetHeader("Content-Disposition", fmt.Sprintf(`attachment;filename="%s"`, filename))
```
The same here too

```suggestion
	return YAML(status, body).
		SetHeader("Content-Type", "application/yaml").
		SetHeader("Content-Disposition", fmt.Sprintf(`attachment;filename="%s"`, filename))
```

JacobsonMT replied:
Fixed b68338c787c7dc807a33b6635c6ad130e7dcf3dd

grobinson-grafana commented on pkg/services/ngalert/api/api_provisioning.go:
```diff
@@ -360,3 +426,26 @@ func determineProvenance(ctx *models.ReqContext) alerting_models.Provenance {
+func exportResponse(c *models.ReqContext, body any) response.Response {
+	format := "json"
+	acceptHeader := c.Req.Header.Get("Accept")
+	if strings.Contains(acceptHeader, "yaml") && !strings.Contains(acceptHeader, "json") {
```
These can be comma separated, although I appreciate we might not want to do full parsing of the `Accept` header https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Accept

JacobsonMT replied:
I'm not sure I follow. If you mean that the Accept header can contain both, then I agree it's up to us to decide what to do here. I chose to only export `yaml` if `json` was not present but I'm open to ideas here.

grobinson-grafana commented on pkg/services/provisioning/alerting/file/rules_types.go:
```diff
@@ -175,3 +173,130 @@ func (queryV1 *QueryV1) mapToModel() (models.AlertQuery, error) {
+	PanelID      int64                      `json:"panelId,omitempty" yaml:"panelId,omitempty"`
+	NoDataState  models.NoDataState         `json:"noDataState" yaml:"noDataState"`
+	ExecErrState models.ExecutionErrorState `json:"execErrState" yaml:"execErrState"`
+	For          model.Duration             `json:"for" yaml:"for"`
```
@alexmobo we will need to update this when we add `ForError`

JohnnyQQQQ reviewed (changes requested):
LGTM, there is one change that I'm missing and it's about the documentation. We should update the documentation [here](https://github.com/grafana/grafana/blob/db51e963de5cd362cd52d24462b33c237c8f8514/docs/sources/alerting/set-up/provision-alerting-resources/file-provisioning/index.md#L32) to use the new endpoint.

JacobsonMT committed:
Improve indenting response.go

JacobsonMT committed:
Regenerate api

JacobsonMT committed:
Regenerate provisioning API docs markdown

JacobsonMT committed:
Update docs to use export endpoints during setup

JacobsonMT commented:
@JohnnyQQQQ

>LGTM, there is one change that I'm missing and it's about the documentation. We should update the documentation [here](https://github.com/grafana/grafana/blob/db51e963de5cd362cd52d24462b33c237c8f8514/docs/sources/alerting/set-up/provision-alerting-resources/file-provisioning/index.md#L32) to use the new endpoint.

Updated api markdown and docs to use export endpoints here: https://github.com/grafana/grafana/commit/ac1a3f1be0176d35569c4be33e93b255789221c4 and https://github.com/grafana/grafana/commit/df5b5a242aa31fbb2481485a4b3feed1d4a55335

brendamuir reviewed (approved):
Looks good from the docs perspective! thanks!

JacobsonMT committed:
Fix conflicts from rebase and regenerate api

JohnnyQQQQ reviewed (approved):
LGTM

JacobsonMT committed:
Fix rebase conflicts and regenerate api
